index.ts

```diff
@@ -0,0 +1,1946 @@
+          .expire(`${redisPreKey}:${this.serviceName}:${this.instanceID}:health:log`, ONE_WEEK_IN_SECONDS)
+          .exec();
+
+        if (this.mcMessageChannelClient) {
+          promises.push(this.mcMessageChannelClient.quitAsync());
+        }
+        if (this.mcDirectMessageChannelClient) {
+          promises.push(this.mcDirectMessageChannelClient.quitAsync());
+        }
+      }
+      Object.keys(this.messageChannelPool).forEach((keyname) => {
+        promises.push(this.messageChannelPool[keyname].quitAsync());
+      });
+      if (this.redisdb) {
+        this.redisdb.del(`${redisPreKey}:${this.serviceName}:${this.instanceID}:presence`, () => {
+          this.redisdb.quit();
+          Promise.all(promises).then(resolve);
+        });
+        this.redisdb.quit();
+        Promise.all(promises).then(resolve);
+      } else {
+        Promise.all(promises).then(resolve);
+      }
+      this.initialized = false;
+      this.instanceID = INSTANCE_ID_NOT_SET;
+    });
+  }
+
+  /**
+   * @name _connectToRedis
+   * @summary Configure access to Redis and monitor emitted events.
+   * @private
+   * @param {object} config - Redis client configuration
+   * @return {object} promise - resolves or reject
+   */
+  _connectToRedis(config: Types.IRedisOptions): Promise<Types.IRedis> {
+    let retryStrategy = config.retry_strategy;
+    delete config.retry_strategy;
+    let redisConnection = new RedisConnection(config, 0, this.testMode);
+    HYDRA_REDIS_DB = redisConnection.redisConfig.db;
+    return redisConnection.connect(retryStrategy)
+      .then((client) => {
+        this.redisdb = client;
+        client
+          .on('reconnecting', () => {
+            this._logMessage('error', 'Reconnecting to Redis server...');
+          })
+          .on('warning', (warning) => {
+            this._logMessage('error', `Redis warning: ${warning}`);
+          })
+          .on('end', () => {
+            this._logMessage('error', 'Established Redis server connection has closed');
+          })
+          .on('error', (err) => {
+            this._logMessage('error', `Redis error: ${err}`);
+          });
+        return client;
+      });
+  }
+
+  /**
+   * @name _getKeys
+   * @summary Retrieves a list of Redis keys based on pattern.
+   * @param {string} pattern - pattern to filter with
+   * @return {object} promise - promise resolving to array of keys or or empty array
+   */
+  _getKeys(pattern: string): Promise<Array<string>> {
+    return new Promise((resolve, _reject) => {
+      if (this.testMode) {
+        this.redisdb.keys(pattern, (err, result) => {
+          if (err) {
+            resolve([]);
+          } else {
+            resolve(result);
+          }
+        });
+      } else {
+        let doScan = (cursor, pattern, retSet) => {
+          this.redisdb.scan(cursor, 'MATCH', pattern, 'COUNT', KEYS_PER_SCAN, (err, result) => {
```

Comment:
the interface to redis should return promises and then just use async/await for cleaner code.

Reply:
Yeah, the interface between `node_redis` and `ioredis` is slightly different - I'd like to abstract that away as you described.